### PR TITLE
ID-698-Sam-GET-user-endpoint

### PR DESCRIPTION
**Pull request by sjkobori**
Ticket: https://broadworkbench.atlassian.net/browse/ID-698

What:

This PR adds a GET user admin endpoint so that it is possible to retrieve a user record from sam.

Why:

This will allow easier access to sam user data which previously was accessed though the database directly.

How:

This PR adds a v2 user admin route which returns a user record

---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes

### CODE REVIEW

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/AdminUserRoutesSpec.scala [hunk 1]
```diff
@@ -325,27 +306,78 @@ class AdminUserRoutesSpec extends AnyFlatSpec with Matchers with ScalatestRouteT
...
+      .withNonAdminUser()
+      .withEnabledUser(defaultUser) // "persisted/enabled" user we will check the status of
       .build
     // Act
     Delete(s"/api/admin/v1/user/$defaultUserId/petServiceAccount/myproject") ~> samRoutes.route ~> check {
       status shouldEqual StatusCodes.Forbidden
     }
   }
+
+  "GET /admin/v2/user/{userId}" should "get the corresponding user object (as an admin)" in {
```
Comment (Shakespeared) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/AdminUserRoutesSpec.scala]: Nit: maybe reword this slightly to something like "get the corresponding user object (if the requesting user is an admin)", just to make it extra explicit that it's the user calling the endpoint who has to be an admin

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/AdminUserRoutesSpec.scala [hunk 2]
```diff
@@ -325,27 +306,78 @@ class AdminUserRoutesSpec extends AnyFlatSpec with Matchers with ScalatestRouteT
...
+    Get(s"/api/admin/v2/user/$defaultUserId") ~> samRoutes.route ~> check {
+      status shouldEqual StatusCodes.OK
+      responseAs[SamUser] shouldEqual defaultUser
+    }
+  }
+
+  it should "not allow access to a user object (as an non admin)" in {
+    // Arrange
+    val samRoutes = new MockSamRoutesBuilder(allUsersGroup)
+      .withEnabledUser(defaultUser) // "persisted/enabled" user we will check the status of
```
Comment (Shakespeared) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/AdminUserRoutesSpec.scala]: Nit: do you mind putting these in the same order as the previous test case? e.g. always having the user calling the endpoint first and the user record being retrieved second. It just makes it easier to read through the tests quickly

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/api/AdminUserRoutesSpec.scala [hunk 3]
```diff
@@ -325,27 +306,78 @@ class AdminUserRoutesSpec extends AnyFlatSpec with Matchers with ScalatestRouteT
...
+      .withNonAdminUser()
+      .build
+
+    // Act and Assert
+    Get(s"/api/admin/v2/user/$defaultUserId") ~> samRoutes.route ~> check {
+      status shouldEqual StatusCodes.Forbidden
+    }
+  }
+
+  it should "not allow access even if a user object does not exist (as an admin)" in {
```
Comment (Shakespeared) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/api/AdminUserRoutesSpec.scala]: Instead of me micromanaging your test names, do you mind taking one last look through and making sure the naming of tests and test setup is consistent? I think it's fine as it is but could be even better :) For instance little things like "return a user object" or something instead of just "allow access," and trying to make the admin/non admin distinction easy to parse at a glance. Overall I think these tests look good though!

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/service/UserServiceSpecs/GetUserSpec.scala [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+          .withEnabledUser(userWithBothIds)
+          .build
+        // Act
+        val response = runAndWait(userService.getUser(userWithBothIds.id, samRequestContext))
+
+        // Assert
+        assert(response.nonEmpty, "Getting a user who exists should return a user object object")
+        assert(
+          response.get equals userWithBothIds,
+          "Getting a user with who exists should return the corresponding user."
```
Comment (Shakespeared) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/service/UserServiceSpecs/GetUserSpec.scala]: I think "user with who exists" is a typo

**Review by Shakespeared — APPROVED** [review]
overall looks great! I had a few nitpicks, mostly just to improve some of the test naming to make it super clear and readable.

File: src/main/resources/swagger/api-docs.yaml [hunk 5]
```diff
@@ -3633,14 +3665,66 @@ components:
...
+      type: object
+      properties:
+        googleSubjectId:
+          type: string
+          description: User's Google subject Id
+        email:
+          type: string
+          description: User's email address
+          format: email
+        azureB2CId:
```
Comment (tlangs) [inline src/main/resources/swagger/api-docs.yaml]: Are we sure we want the B2C ID to be updatable?

File: src/main/resources/swagger/api-docs.yaml [hunk 6]
```diff
@@ -3633,14 +3665,66 @@ components:
...
+          type: string
+          description: User's Azure B2C Id
+        enabled:
+          type: boolean
+          description: Whether or not the user is enabled
+        acceptedTosVersion:
```
Comment (tlangs) [inline src/main/resources/swagger/api-docs.yaml]: The accepted TOS version should only be updated via the TOS endpoints.

File: src/main/resources/swagger/api-docs.yaml (hunk 5 above)
Comment (sjkobori) [reply src/main/resources/swagger/api-docs.yaml]: As it stands right now, the only field that is updatable in practice is the email. The UpdateUserRequest object does contain all of these fields and I realized that the swagger doc was not in sync with the UpdateUser endpoint's request object. We still need to make decisions on what fields we are allowing to update. I think ultimately that the B2CId might be useful to update without having to hit the db directly. We can restrict updating this field to only superadmins or something. Right now, we are using the sam user directives to manage authorization and the level needed to use this endpoint are admins. Going forward, I think we will have a permissions service that had handle what actions are allowed by a user more granularly (nobody can update a user email except for themselves and users with admin roles)

File: src/main/resources/swagger/api-docs.yaml (hunk 6 above)
Comment (sjkobori) [reply src/main/resources/swagger/api-docs.yaml]: This is another thing I wanted to talk about. I do agree that I think only a user can accept their tos (although I think it might be valuable to edit the version if for some reason there is a data discrepancy), but I think eventually we would want to do away with the specific endpoints for updating one field of a user. For now, I can take this out of the swagger doc, would you also like me to remove it from the UpdateUserRequest object in code (right now it is not doing anything) or just leave that as it is?

Comment (tlangs) [reply src/main/resources/swagger/api-docs.yaml]: The Update model should probably only contain update-able fields, since our API definition should reflect the functionality of the endpoint. As for the B2C ID, I we need to do a lot of hard thinking about if that should be update-able. I don't think we've gotten any user requests or support tickets involving updating the B2C ID for a user. We do know that there's a need to update emails and Google Subject IDs, so that's probably what we should focus on and provide for now.

Comment (tlangs) [reply src/main/resources/swagger/api-docs.yaml]: But, to another point, a set of CRUD endpoints should probably use the same object for requests and responses. So, if we were to have just a `SamUser` object with all the user fields, and use that as both the request and response, then we could say that the only fields that are update-able are the email and Google Subject ID.

Comment (sjkobori) [reply src/main/resources/swagger/api-docs.yaml]: I think that we should not be using the same request and response objects since it makes it difficult to know what is required in a request vs what you know is in the response (that is what made updating the `PolicyIdentifiers `complex in the first place). So for example, in the `GET` user endpoint, we can say that the email is a required field. But in the `PATCH` user endpoint, email is not required. This could be done in the `POST` user endpoint by marking certain required field as read only since you need a complete object to make this request. So I still think that there should be a `UpdateUserRequest` object, but I will update it to only have the fields we currently support updating.

**Review by Ghost-in-a-Jar — CHANGES_REQUESTED** [review]
User update request should only have googleSubjectId, email, enabled, and you should wire those to be actually updated by admins if included in the request.

There are probably additional steps needed when updating the email, besides just updating the db record. (thurloe?)

We may want to consider implementing email in a separate pr.

**Comment by sjkobori** [comment]
> User update request should only have googleSubjectId, email, enabled, and you should wire those to be actually updated by admins if included in the request.
> 
> There are probably additional steps needed when updating the email, besides just updating the db record. (thurloe?)
> 
> We may want to consider implementing email in a separate pr.

I think since now the PATCH user endpoint has been disabled in prod, I would like to keep it that way for this ticket and add all the intended features for that endpoint in ticket: https://broadworkbench.atlassian.net/browse/ID-718 . This way it is not longer blocking people that need the GET user endpoint.

**Review by Ghost-in-a-Jar — APPROVED** [review]
